### ISSUE TITLE
ActiveSync: Calendar: Error Sync status 3

### ISSUE THREAD

**Issue by benbucksch** [mustang-im/mustang#282]
### Reproduction
* Randomly clicking around on folders in the ActiveSync account

### Actual result
* Error: ActiveSync Sync status 3
* Stack:
![image](https://github.com/user-attachments/assets/ff4dc00c-3b6b-4663-bb79-8154e9e648a3)
* Note that the stack shows the calendar

**Comment by benbucksch** [comment]
@NeilRashbrook Can you please:
* Handle the errors that we can recover from? E.g. this one should automatically trigger the re-sync, as requested, and not throw an error.
* Show human-readable, translated error messages for all other sync codes that we cannot handle?

**Comment by NeilRashbrook** [comment]
What do you want to do in the case of an edit operation? Doing a re-sync will throw away all our data and refresh it from the server.

**Comment by benbucksch** [comment]
Keep the edit in a cache, resync, then re-apply the chnage on top of the newly synced data.

**Comment by NeilRashbrook** [comment]
If you're getting a lot of sync status 3, it probably means you have duplicate entries in your calendar database. I noticed I had two sets of ActiveSync calendars and when I deleted one set the sync status 3 errors went away.

**Comment by NeilRashbrook** [comment]
One of my calendars duplicated itself again; I'm not sure how that would happen, the code checks `appGlobal.calendars` for a matching calendar first. Strange also that it was only one calendar that got duplicated.

**Comment by NeilRashbrook** [comment]
> Keep the edit in a cache, resync, then re-apply the chnage on top of the newly synced data.

I've just realised that this doesn't make sense because mustang doesn't actually keep track of what the change was. Anyway, I've created branch `neil/activesync-sync-status-3` which checks each `Sync` command for a status code of `3`, and if so, first resyncs, then checks whether the item being edited still exists, and if so, retries the operation. (I special-case a delete operation so that it does not fail if the item has already been deleted on the server.)

**Comment by benbucksch** [comment]
Yes, this is pretty much what I had in mind.

I see what you do with the makeSyncRequest in ActiveSyncEMail class. Here's abother idea with less changes:
* Do the catch error 3 and resync in folder.
* Add a "retryOnlyIf" function param to makeSyncRequest
* Let callers in ActiveSyncEMail pass that for deleted or not deleted checks

Other point: You do addPingable() on every call to sync(). Do the pingables accumulate or does the callee check for already existing ones?

**Comment by NeilRashbrook** [comment]
Pingables don't accumulate but they do become recent so when the server says that it wants you to ping fewer folders it will keep the ones you most recently accessed.

**Comment by NeilRashbrook** [comment]
Surely the problem with calling `syncMessages` from inside `makeSyncRequest` is that you're inside the lock at that point?

**Comment by NeilRashbrook** [comment]
One approach might be to always list messages first before performing any other operation, as that means that you should only need to handle sync status 3 in one place.

**Comment by benbucksch** [comment]
Sorry, I think you misunderstood me. I only meant to merge the new code 
into the existing `ActiveAyncFolder.makeSyncRequest()` and avoid to add 
the new `ActiveSyncEMail.makeSyncRequest()` etc.. The operation-specific 
check before the retry (is already deleted, is not deleted etc.) could be 
a closure function, passed in as optional parameter.

**Comment by NeilRashbrook** [comment]
Yes, that's exactly what my problem is; `ActiveSyncEMail.makeSyncRequest` needs to make three calls to `ActiveSyncFolder.makeSyncRequest`.

**Comment by benbucksch** [comment]
I see

**Comment by benbucksch** [comment]
See #331 for a possible cause

**Comment by NeilRashbrook** [comment]
Ugh, my calendars are still duplicating themselves, resulting in sync status 3... I really need to find out why that happens.
